File: app/src/main/java/dev/n3shemmy3/coffre/ui/screen/settings/SettingsScreen.java
package dev.n3shemmy3.coffre.ui.screen.settings;
/*
 * Copyright (C) 2025 Shemmy
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import android.content.Intent;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Build;
import android.os.Bundle;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.PreferenceScreen;
import androidx.preference.PreferenceViewHolder;
import androidx.recyclerview.widget.RecyclerView;

import com.google.gson.Gson;

import dev.n3shemmy3.coffre.R;
import dev.n3shemmy3.coffre.backend.entity.Profile;
import dev.n3shemmy3.coffre.ui.adapter.PreferencesAdapter;
import dev.n3shemmy3.coffre.ui.base.BaseScreen;
import dev.n3shemmy3.coffre.ui.item.decorator.VerticalSpaceItemDecoration;
import dev.n3shemmy3.coffre.ui.navigator.Navigator;
import dev.n3shemmy3.coffre.ui.screen.setup.SetupScreen;
import dev.n3shemmy3.coffre.ui.utils.FileUtils;
import dev.n3shemmy3.coffre.ui.utils.PrefUtil;

public class SettingsScreen extends BaseScreen {
    @Override
    protected int getLayoutResId() {
        return R.layout.screen_settings;
    }

    @RequiresApi(api = Build.VERSION_CODES.UPSIDE_DOWN_CAKE)
    @Override
    protected void onCreateScreen(View root, Bundle state) {
        super.onCreateScreen(root, state);
        root.findViewById(R.id.settingsContainer).setClipToOutline(true);
        root.findViewById(R.id.item).setOnClickListener(view -> Navigator.push(getScreenManager(), new SetupScreen()));
        ImageView itemStartIcon = root.findViewById(R.id.itemStartIcon);
        TextView itemTitle = root.findViewById(R.id.itemTitle);
        Profile profile = new Gson().fromJson(PrefUtil.getString(Profile.key), Profile.class);
        if (profile != null) {
            itemStartIcon.setImageBitmap(FileUtils.retrieveImageFromPrivateStorage(requireActivity(), profile.getAvatar()));
            itemTitle.setText(profile.getName());
        }

    }

    @RequiresApi(api = Build.VERSION_CODES.UPSIDE_DOWN_CAKE)
    @Override
    public void onResume() {
        super.onResume();

    }

    public static class PreferencesScreen extends PreferenceFragmentCompat {
        @Override
        public void onCreatePreferences(@Nullable Bundle savedInstanceState, @Nullable String rootKey) {
            setPreferencesFromResource(R.xml.settings, rootKey);

        }

        @Override
        public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
            super.onViewCreated(view, savedInstanceState);
            setDivider(new ColorDrawable(Color.TRANSPARENT));
            setDividerHeight(0);
            RecyclerView recyclerView = getListView();
            if (recyclerView != null) {
                recyclerView.setClipToPadding(false);
                recyclerView.setClipToOutline(true);
                recyclerView.setOverScrollMode(View.OVER_SCROLL_NEVER);
                recyclerView.addItemDecoration(new VerticalSpaceItemDecoration(4));
            }
        }

        @NonNull
        @Override
        protected RecyclerView.Adapter<PreferenceViewHolder> onCreateAdapter(@NonNull PreferenceScreen preferenceScreen) {
            PreferencesAdapter adapter = new PreferencesAdapter(preferenceScreen);
            adapter.notifyItemRangeChanged(0, adapter.getItemCount());
            return adapter;
        }

        @Override
        public boolean onPreferenceTreeClick(@NonNull Preference preference) {
            switch (preference.getKey()) {
                case "looknfeel": {
                    Navigator.push(requireActivity().getSupportFragmentManager(), new LookNFeelSettingsScreen());
                }
                break;
                case "formats": {
                    Navigator.push(requireActivity().getSupportFragmentManager(), new FormatSettingsScreen());
                }
                break;
                case "about": {
                    Navigator.push(requireActivity().getSupportFragmentManager(), new AboutSettingsScreen());
                }
                break;
            }
            return super.onPreferenceTreeClick(preference);
        }
    }
}

File: app/src/main/kotlin/dev/n3shemmy3/coffre/ui/screen/settings/SettingsScreen.kt
/*
 *
 *  * Copyright (C) 2025 Shemmy
 *  *
 *  * This program is free software: you can redistribute it and/or modify
 *  * it under the terms of the GNU General Public License as published by
 *  * the Free Software Foundation, version 3 of the License.
 *  *
 *  * This program is distributed in the hope that it will be useful,
 *  * but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  * GNU General Public License for more details.
 *  *
 *  * You should have received a copy of the GNU General Public License
 *  * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */

package dev.n3shemmy3.coffre.ui.screen.settings

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Info
import androidx.compose.material.icons.outlined.Palette
import androidx.compose.material.icons.outlined.Person4
import androidx.compose.material.icons.outlined.SettingsBackupRestore
import androidx.compose.material.icons.outlined.Tag
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.lifecycle.viewmodel.compose.viewModel
import dev.n3shemmy3.coffre.R
import dev.n3shemmy3.coffre.data.action.Action
import dev.n3shemmy3.coffre.data.entity.Preference
import dev.n3shemmy3.coffre.data.viewmodel.MainViewModel
import dev.n3shemmy3.coffre.ui.component.PreferenceItem
import dev.n3shemmy3.coffre.ui.component.PreferenceTitle
import dev.n3shemmy3.coffre.ui.navigation.Route
import dev.n3shemmy3.coffre.util.getItemShape

@Composable
fun SettingsScreen(viewModel: MainViewModel) {

    BaseSettingsScreenContent(
        stringResource(R.string.screen_settings),
        onBackClicked = {
            viewModel.onAction(Action.ViewFlow.Close(Route.SETTINGS))
        }
    ) {
        Column(
            verticalArrangement = Arrangement.spacedBy(4.dp)
        ) {
            PreferenceTitle(stringResource(R.string.profile))
            PreferenceItem(
                icon = Icons.Outlined.Person4,
                title = stringResource(R.string.preference_profile),
                summary = stringResource(R.string.summary_profile),
                onClick = {},
                shape = CircleShape
            )
        }

        Column(
            verticalArrangement = Arrangement.spacedBy(4.dp)
        ) {
            PreferenceTitle(stringResource(R.string.screen_settings))
            val preferences = remember {
                listOf(
                    Preference(
                        Icons.Outlined.Palette,
                        R.string.preference_look_n_feel,
                        R.string.summary_look_and_feel,
                        onClick = {
                            viewModel.onAction(Action.ViewFlow.Open(Route.LOOKNFEELSETTINGS))
                        }
                    ),
                    Preference(
                        Icons.Outlined.Tag,
                        R.string.preference_format,
                        R.string.summary_format,
                        onClick = {
                            viewModel.onAction(Action.ViewFlow.Open(Route.FORMATSETTINGS))
                        }
                    ),
                    Preference(
                        Icons.Outlined.SettingsBackupRestore,
                        R.string.preference_backup_n_restore,
                        R.string.summary_backup_n_restore,
                        onClick = {
                            viewModel.onAction(Action.ViewFlow.Open(Route.BACKUPNRESTORESETTINGS))
                        }
                    ),
                    Preference(
                        Icons.Outlined.Info,
                        R.string.preference_about,
                        R.string.summary_about,
                        onClick = {
                            viewModel.onAction(Action.ViewFlow.Open(Route.ABOUTSETTINGS))
                        }
                    )
                )
            }

            preferences.forEachIndexed { index, item ->
                PreferenceItem(
                    item.icon,
                    stringResource(item.title),
                    stringResource(item.summary),
                    onClick = item.onClick,
                    shape = getItemShape(index, preferences.lastIndex)
                )
            }
        }

    }
}

@Preview
@Composable
fun SettingsScreenPreview() {
    SettingsScreen(viewModel = viewModel())
}